api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -827,21 +815,185 @@ private String generateRandomEightCharacterQualifier() {
-                        dataSetExportRequest.getKernelType()))
+    return Stream.concat(
+            queriesByDomain.entrySet().stream()
+                .filter(
+                    query ->
+                        !Domain.WHOLE_GENOME_VARIANT
+                            .toString()
+                            .equals(query.getKey())) // This filter can be removed once valid SQL is
+                // generated for WGS
+                .map(
+                    entry ->
+                        prerequisites
+                            + "\n\n"
+                            + generateNotebookUserCode(
+                                entry.getValue(),
+                                Domain.fromValue(entry.getKey()),
+                                dataSetExportRequest.getDataSetRequest().getName(),
+                                dbWorkspace.getCdrVersion().getName(),
+                                qualifier,
+                                dataSetExportRequest.getKernelType())),
+            generateWgsCode(dataSetExportRequest, dbWorkspace, qualifier).stream())
         .collect(Collectors.toList());
   }
 
+  private List<String> generateWgsCode(
+      DataSetExportRequest dataSetExportRequest, DbWorkspace dbWorkspace, String qualifier) {
+    if (!dataSetExportRequest.getGenerateGenomicsAnalysisCode()) {
+      return new ArrayList<>();
+    }
+
+    if (!workbenchConfigProvider.get().featureFlags.enableGenomicExtraction) {
+      throw new NotImplementedException();
+    }
+
+    if (Strings.isNullOrEmpty(dbWorkspace.getCdrVersion().getWgsBigqueryDataset())) {
+      throw new FailedPreconditionException(
+          "The workspace CDR version does not have whole genome data");
+    }
+
+    if (dataSetExportRequest.getKernelType().equals(KernelTypeEnum.R)) {
+      return generateGenomicsAnalysisComment_R();
+    }
+
+    switch (dataSetExportRequest.getGenomicsAnalysisTool()) {
+      case HAIL:
+        return generateHailCode(qualifier, dataSetExportRequest);
+      case PLINK:
+        return generatePlinkCode(qualifier, dataSetExportRequest);
+      case NONE:
+        return generateDownloadVcfCode(qualifier, dataSetExportRequest);
+      default:
+        throw new BadRequestException("Invalid Genomics Analysis Tool");
+    }
+  }
+
+  private Optional<String> getExtractionDirectory(Long datasetId) {
+    try {
+      return Optional.of(
+              submissionDao
+                  .findMostRecentValidExtractionByDataset(dataSetDao.findById(datasetId).get())
+                  .get()
+                  .getOutputDir()
+                  .replaceFirst("/$", "")) // Drop trailing slash if exists
+          .filter(dir -> !dir.isEmpty());
+    } catch (NoSuchElementException e) {
+      return Optional.empty();
+    }
+  }
+
+  private String generateVcfDirEnvName(String qualifier) {
+    return "DATASET_" + qualifier + "_VCF_DIR";
+  }
+
+  private static final String MISSING_EXTRACTION_DIR_PLACEHOLDER =
```

Comment:
nit: convention is generally to put constants at the top of the file in Java (maybe with some rare exceptions)